test/core/tsi/spiffe_ssl_transport_security_test.cc

```diff
@@ -0,0 +1,438 @@
+#include "gtest/gtest.h"
+#include "src/core/credentials/transport/security_connector.h"
+#include "src/core/tsi/ssl_transport_security.h"
+#include "src/core/tsi/transport_security.h"
+#include "src/core/tsi/transport_security_interface.h"
+#include "src/core/util/crash.h"
+#include "test/core/test_util/test_config.h"
+#include "test/core/test_util/tls_utils.h"
+#include "test/core/tsi/transport_security_test_lib.h"
+
+extern "C" {
+#include <openssl/crypto.h>
+#include <openssl/pem.h>
+}
+
+namespace {
+constexpr absl::string_view kCaPemPath =
+    "test/core/tsi/test_creds/spiffe_end2end/ca.pem";
+constexpr absl::string_view kClientKeyPath =
+    "test/core/tsi/test_creds/spiffe_end2end/client.key";
+constexpr absl::string_view kClientCertPath =
+    "test/core/tsi/test_creds/spiffe_end2end/client_spiffe.pem";
+constexpr absl::string_view kServerKeyPath =
+    "test/core/tsi/test_creds/spiffe_end2end/server.key";
+constexpr absl::string_view kServerCertPath =
+    "test/core/tsi/test_creds/spiffe_end2end/server_spiffe.pem";
+constexpr absl::string_view kServerChainKeyPath =
+    "test/core/tsi/test_creds/spiffe_end2end/leaf_signed_by_intermediate.key";
+constexpr absl::string_view kServerChainCertPath =
+    "test/core/tsi/test_creds/spiffe_end2end/leaf_and_intermediate_chain.pem";
+constexpr absl::string_view kClientSpiffeBundleMapPath =
+    "test/core/tsi/test_creds/spiffe_end2end/client_spiffebundle.json";
+constexpr absl::string_view kServerSpiffeBundleMapPath =
+    "test/core/tsi/test_creds/spiffe_end2end/server_spiffebundle.json";
+
+constexpr absl::string_view kNonSpiffeKeyPath =
+    "test/core/tsi/test_creds/crl_data/valid.key";
+constexpr absl::string_view kNonSpiffeCertPath =
+    "test/core/tsi/test_creds/crl_data/valid.pem";
+constexpr absl::string_view kNonSpiffeCAPath =
+    "test/core/tsi/test_creds/crl_data/ca.pem";
+constexpr absl::string_view kMultiSanKeyPath =
+    "test/core/tsi/test_creds/spiffe_end2end/multi_san.key";
+constexpr absl::string_view kMultiSanCertPath =
+    "test/core/tsi/test_creds/spiffe_end2end/multi_san_spiffe.pem";
+constexpr absl::string_view kInvalidUtf8SanKeyPath =
+    "test/core/tsi/test_creds/spiffe_end2end/invalid_utf8_san.key";
+constexpr absl::string_view kInvalidUtf8SanCertPath =
+    "test/core/tsi/test_creds/spiffe_end2end/invalid_utf8_san_spiffe.pem";
+
+class SpiffeSslTransportSecurityTest
+    : public testing::TestWithParam<tsi_tls_version> {
+ protected:
+  // A tsi_test_fixture implementation.
+  class SslTsiTestFixture {
+   public:
+    SslTsiTestFixture(absl::string_view server_key_path,
+                      absl::string_view server_cert_path,
+                      absl::string_view client_key_path,
+                      absl::string_view client_cert_path,
+                      absl::string_view server_spiffe_bundle_map_path,
+                      absl::string_view client_spiffe_bundle_map_path,
+                      absl::string_view ca_path, bool expect_server_success,
```

Comment:
Strongly suggest making `ca_path` an optional parameter and only populating `root_cert_` if it is not `std::nullopt`.

Rationale: This way, we can have more confidence that verification is passing because the SPIFFE verification is working correctly, rather than somehow silently failing back to the flat trust bundle.